balena/utils.py

```diff
@@ -46,3 +56,182 @@ def compare(a, b):
+                },
+            },
+            "$expand": {
+                "image": {
+                    "$select": ["id"],
+                    "$expand": {
+                        "is_a_build_of__service": {
+                            "$select": ["id", "service_name"],
+                        },
+                    },
+                },
+                **(
+                    {
+                        "is_provided_by__release": {
+                            "$select": ["id", "commit", "raw_version"],
+                        },
+                    }
+                    if expand_release
+                    else {}
+                ),
+            },
+        },
+    }
+
+
+def get_single_install_summary(raw_data: Any) -> Any:
+    # TODO: Please compare me to node-sdk version
+    """
+    Builds summary data for an image install or gateway download
+
+    """
+
+    image = raw_data["image"][0]
+    service = image["is_a_build_of__service"][0]
+    release = None
+
+    if "is_provided_by__release" in raw_data:
+        release = raw_data["is_provided_by__release"][0]
+
+    install = {
+        "service_name": service["service_name"],
+        "image_id": image["id"],
+        "service_id": service["id"],
+    }
+
+    if release:
+        install["commit"] = release["commit"]
+
+    raw_data.pop("is_provided_by__release", None)
+    raw_data.pop("image", None)
+    install.update(raw_data)
+
+    return install
+
+
+def generate_current_service_details(raw_device: Any) -> Any:
+    # TODO: Please compare me to node-sdk version
+    groupedServices = defaultdict(list)
+
+    for obj in [
+        get_single_install_summary(i)
+        for i in raw_device.get("image_install", [])
+    ]:
+        groupedServices[obj.pop("service_name", None)].append(obj)
+
+    raw_device["current_services"] = dict(groupedServices)
+    raw_device["current_gateway_downloads"] = [
+        get_single_install_summary(i)
+        for i in raw_device.get("gateway_download", [])
+    ]
+    raw_device.pop("image_install", None)
+    raw_device.pop("gateway_download", None)
+
+    return raw_device
+
+
+def is_provisioned(device: Any) -> bool:
+    return (
+        device.get("supervisor_version") is not None
+        and len(device.get("supervisor_version")) > 0
+        and device.get("last_connectivity_event") is not None
+    )
+
+
+def normalize_device_os_version(device: Any) -> Any:
+    if (
+        device.get("os_version") is not None
+        and len(device.get("os_version")) == 0
+        and is_provisioned(device)
+    ):
+        device["os_version"] = "Resin OS 1.0.0-pre"
+
+    return device
+
+
+T = TypeVar("T")
+
+
+def with_supervisor_locked_error(fn: Callable[[], T]) -> T:
+    try:
+        return fn()
+    except RequestError as e:
+        if e.status_code == SUPERVISOR_LOCKED_STATUS_CODE:
+            raise SupervisorLocked()
+        raise e
+
+
+def normalize_balena_semver(os_version):
```

Comment:
AFAICT this has been copy-pasted from the `balena/models/device_os.py` file. Should the original function be deleted?

Reply:
Also, it's the only function w/o type hints in the file. Would be awesome to add it for the sake of consistency.

Reply:
Yep, I am still refactoring device_os local and the types and deduplication is coming on the next commit :)